Topics/System_Design/Authentication_Service_Design.md

```diff
@@ -0,0 +1,86 @@
+
+If you've ever heard of JSON Web Tokens (JWTs), that's a common form of token-based authentication! JWTs embed user data inside the token, allowing validation without database queries, making them suitable for serverless and stateless applications.
+
+JWTs have several advantages:
+
+1.  Compact size for quick transmission.
+2.  Enhanced security through asymmetric encryption.
+3.  Ubiquitous use due to the prevalence of JSON objects.
+4.  Transparency in verifying senders and preventing tampering.
+
+They also have drawbacks; for example, they make it harder to revoke access from the server side, leading to shorter expiration times to mitigate security risks; this can be annoying for a user.
+
+### Designing Token-Based Authentication <a name="token-design"></a>
+There's a few best practices to follow when we're building an authentication service using a token-based service:
+1.  **Use Industry Standards:** Leverage established standards for token creation, such as OAuth 2.0 and OpenID Connect. Following widely accepted standards enhances interoperability and security.
+2.  **Secure Token Storage:** Ensure secure storage of tokens on the client side. Avoid storing sensitive information in local storage or cookies susceptible to attacks. Use HTTP-only cookies for enhanced security.
+3.  **Token Revocation:** Implement mechanisms for token revocation in case of security incidents or when a user logs out. Maintain a blacklist or use token introspection to check the validity of tokens.
+4.  **Token Scoping:** Include a scope or permissions within the token to define the level of access granted to the client, and make sure to give the minimum possible amount of privileges.
+5.  **Token Encryption:** When sensitive information is embedded in tokens, consider encrypting the token content. This adds an extra layer of security, especially for tokens containing user-specific data.
+6.  **Implement Rate Limiting:** Protect against abuse by implementing rate limiting on token-related endpoints. This helps mitigate the risk of brute force attacks or other malicious activities.
+
+## Design Considerations <a name="auth-design"></a>
+
+When it comes to building an authentication service for any kind of digital service, particularly in the context of microservice architectures, the best approach is to "use an API gateway—this is a service deployed in front of the microservices application, which serves as a single endpoint for all user requests." ([Source](https://frontegg.com/blog/authentication)) A good example of this are [interceptors in Axios](https://axios-http.com/docs/interceptors). Every API call in an mobile app or web service will pass through this checkpoint; it'll do something like attach a token, ensure a token is valid, or catch unauthorized access attempts. Other services sending requests to this checkpoint won't be aware of the underlying service at this gateway; it'll just package the requests accordingly, authenticate, and forward them. 
```

Comment:
Thank you for adding the source in such an easy link! Provides me a way of learning where the quote came from